File: app/src/main/java/de/fhswf/fahrradverleih/android/model/Ausleihe.java
package de.fhswf.fahrradverleih.android.model;

import android.os.Parcelable;

import androidx.annotation.NonNull;

import java.io.Serializable;
import java.time.LocalDateTime;

public class Ausleihe implements Serializable {

    @NonNull
    private String id;

    @NonNull
    private Fahrrad fahrrad;

    @NonNull
    private TarifT tarif;

    @NonNull
    private LocalDateTime von;

    @NonNull
    private LocalDateTime bis;

    public Ausleihe(@NonNull String id, @NonNull Fahrrad fahrrad, @NonNull TarifT tarif,
                    @NonNull LocalDateTime von, @NonNull LocalDateTime bis) {
        this.id = id;
        this.fahrrad = fahrrad;
        this.tarif = tarif;
        this.von = von;
        this.bis = bis;
    }

    public Ausleihe() {
        this("", new Fahrrad(), new TarifT(), LocalDateTime.now(), LocalDateTime.now());
    }

    @NonNull
    public String getId() {
        return id;
    }

    public void setId(@NonNull String id) {
        this.id = id;
    }

    @NonNull
    public Fahrrad getFahrrad() {
        return fahrrad;
    }

    public void setFahrrad(@NonNull Fahrrad fahrrad) {
        this.fahrrad = fahrrad;
    }

    @NonNull
    public TarifT getTarif() {
        return tarif;
    }

    public void setTarif(@NonNull TarifT tarif) {
        this.tarif = tarif;
    }

    @NonNull
    public LocalDateTime getVon() {
        return von;
    }

    public void setVon(@NonNull LocalDateTime von) {
        this.von = von;
    }

    @NonNull
    public LocalDateTime getBis() {
        return bis;
    }

    public void setBis(@NonNull LocalDateTime bis) {
        this.bis = bis;
    }

    public boolean isAktiv() {
        return bis.isAfter(LocalDateTime.now());
    }
}

File: app/src/main/java/de/fhswf/fahrradverleih/android/model/Ausleihe.kt
package de.fhswf.fahrradverleih.android.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

@Serializable
data class Ausleihe(
    val id: String = "",
    val fahrrad: Fahrrad = Fahrrad(),
    val tarif: TarifT = TarifT(),
    @Serializable(with = JsDateSerializer::class)
    val von: LocalDateTime = LocalDateTime.now(),
    @Serializable(with = JsDateSerializer::class)
    val bis: LocalDateTime = LocalDateTime.now(),
)

/**
 * Serialisiert die JSON-Felder mit Datum entsprechend des vom Backend verwendeten
 * Formats nach ISO 8601.
 */
object JsDateSerializer : KSerializer<LocalDateTime> {
    override fun deserialize(decoder: Decoder): LocalDateTime =
        LocalDateTime.parse(decoder.decodeString(), DateTimeFormatter.ISO_DATE_TIME)

    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("LocalDateTime", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: LocalDateTime) =
        encoder.encodeString(value.format(DateTimeFormatter.ISO_DATE_TIME))

}
